File: playwright/autofill.js
#!/usr/bin/env node
/**
 * Playwright autofill CLI.
 *
 * Usage:
 *   node autofill.js --url <url> [--profile <key>] [--submit] [--headful] [--channel chrome]
 *
 * Strategy:
 *   1. Inject the core engine into every frame (including cross-origin Stripe iframes —
 *      Playwright bypasses same-origin policy via CDP).
 *   2. Call fill() in every frame. Each frame fills whatever selectors it finds.
 *   3. Top frame submits last.
 */

const fs = require('fs');
const path = require('path');
const { chromium } = require('playwright');

const ROOT = path.resolve(__dirname, '..');
const ENGINE_SRC = fs.readFileSync(path.join(ROOT, 'core/autofill-engine.js'), 'utf8');
const PROFILES_PATH = fs.existsSync(path.join(ROOT, 'profiles.json'))
  ? path.join(ROOT, 'profiles.json')
  : path.join(ROOT, 'profiles.example.json');

function parseArgs(argv) {
  const args = { submit: false, headful: false, waitFor: 3000 };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    const next = () => argv[++i];
    if (a === '--url') args.url = next();
    else if (a === '--profile') args.profile = next();
    else if (a === '--submit') args.submit = true;
    else if (a === '--headful' || a === '--headed') args.headful = true;
    else if (a === '--channel') args.channel = next();
    else if (a === '--wait') args.waitFor = Number(next()) || 3000;
    else if (a === '--help' || a === '-h') args.help = true;
    else console.warn('Unknown arg:', a);
  }
  return args;
}

function help() {
  console.log(`
Auto Fill — Playwright CLI

  --url <url>        Target page URL (required)
  --profile <key>    Profile key from profiles.json (default: defaultProfile)
  --submit           Click submit after filling
  --headful          Show browser window
  --channel chrome   Use installed Chrome instead of bundled Chromium
  --wait <ms>        Wait this long after page load before filling (default: 3000)

Example:
  node autofill.js --url https://example.com/checkout --profile personal --submit --headful
`);
}

async function injectEngineIntoFrame(frame) {
  try {
    // addScriptTag is scoped to each frame; content is the engine IIFE.
    await frame.evaluate(ENGINE_SRC);
    return true;
  } catch (e) {
    // Cross-origin frames that aren't yet loaded will throw; skip them.
    return false;
  }
}

async function fillFrame(frame, profile, submit) {
  try {
    return await frame.evaluate(({ profile, submit }) => {
      if (!window.__AutoFill) return { skipped: true, reason: 'engine missing', url: location.href };
      return {
        url: location.href,
        ...window.__AutoFill.fill(profile, { submit: submit && window.top === window }),
      };
    }, { profile, submit });
  } catch (e) {
    return { error: e.message, url: frame.url() };
  }
}

async function main() {
  const args = parseArgs(process.argv);
  if (args.help || !args.url) { help(); process.exit(args.help ? 0 : 1); }

  const store = JSON.parse(fs.readFileSync(PROFILES_PATH, 'utf8'));
  const profileKey = args.profile || store.defaultProfile;
  const profile = store.profiles[profileKey];
  if (!profile) {
    console.error(`Unknown profile "${profileKey}". Available: ${Object.keys(store.profiles).join(', ')}`);
    process.exit(1);
  }

  console.log(`[autofill] profile="${profileKey}" submit=${args.submit} url=${args.url}`);

  const browser = await chromium.launch({
    headless: !args.headful,
    channel: args.channel,
  });
  const context = await browser.newContext();
  const page = await context.newPage();

  // Pre-inject engine into every new frame the moment it navigates. This catches
  // Stripe iframes that mount after first paint.
  context.on('page', (p) => {
    p.on('frameattached', async (f) => { await injectEngineIntoFrame(f); });
    p.on('framenavigated', async (f) => { await injectEngineIntoFrame(f); });
  });
  page.on('frameattached', async (f) => { await injectEngineIntoFrame(f); });
  page.on('framenavigated', async (f) => { await injectEngineIntoFrame(f); });

  await page.goto(args.url, { waitUntil: 'load' });
  await page.waitForTimeout(args.waitFor);

  // Inject into every frame we can see now.
  for (const f of page.frames()) {
    await injectEngineIntoFrame(f);
  }

  const results = [];
  // Fill each frame independently — they each handle their own input set.
  for (const f of page.frames()) {
    const r = await fillFrame(f, profile, false);
    if (r && (r.filled?.length || r.error || r.skipped)) results.push(r);
  }

  console.log('\n[autofill] per-frame results:');
  for (const r of results) {
    if (r.error) console.log(`  ✗ ${r.url} — ${r.error}`);
    else if (r.skipped) console.log(`  — ${r.url} — ${r.reason}`);
    else console.log(`  ✓ ${r.url}\n    filled: ${r.filled.join(', ') || '(none)'}\n    missed: ${r.missed.join(', ') || '(none)'}`);
  }

  if (args.submit) {
    // Let validation settle after last blur.
    await page.waitForTimeout(300);
    const clicked = await page.evaluate(() => window.__AutoFill?.submitForm());
    console.log(`\n[autofill] submit: ${clicked ? 'clicked' : 'no submit button found in top frame'}`);
    // Give the submission a chance to navigate/respond before we close.
    await page.waitForTimeout(3000);
  } else if (args.headful) {
    console.log('\n[autofill] headful mode — browser left open, Ctrl+C to exit.');
    await new Promise(() => {}); // wait forever
  }

  if (!args.headful) await browser.close();
}

main().catch((e) => { console.error(e); process.exit(1); });

File: python/autofill.py
#!/usr/bin/env python3
"""
Auto Fill — Python Playwright CLI.

Port of playwright/autofill.js. Same flags, same strategy: inject the shared
engine (core/autofill-engine.js) into every frame (including cross-origin
Stripe iframes via CDP), call fill() in each, then submit the top frame.

Usage:
    python autofill.py --url <url> [--profile <key>] [--submit] [--headful]
                       [--channel chrome] [--wait 3000]
"""

from __future__ import annotations

import argparse
import asyncio
import json
import sys
from pathlib import Path

from playwright.async_api import BrowserContext, Frame, Page, async_playwright

ROOT = Path(__file__).resolve().parent.parent
ENGINE_SRC = (ROOT / "core" / "autofill-engine.js").read_text(encoding="utf-8")
PROFILES_PATH = (ROOT / "profiles.json") if (ROOT / "profiles.json").exists() else (ROOT / "profiles.example.json")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        prog="autofill",
        description="Auto-fill payment forms via Playwright (Stripe iframes supported).",
    )
    parser.add_argument("--url", required=True, help="Target page URL")
    parser.add_argument("--profile", help="Profile key from profiles.json (default: defaultProfile)")
    parser.add_argument("--submit", action="store_true", help="Click submit after filling")
    parser.add_argument("--headful", "--headed", action="store_true", help="Show the browser window")
    parser.add_argument("--channel", help="Browser channel (e.g. 'chrome', 'msedge')")
    parser.add_argument("--wait", type=int, default=3000, help="Delay in ms after page load before filling")
    return parser.parse_args()


async def inject_engine(frame: Frame) -> bool:
    """
    Evaluate the engine IIFE in a frame. Returns False if the frame is
    cross-origin and not ready (Playwright still bypasses, but the frame
    might not have a document yet).
    """
    try:
        await frame.evaluate(ENGINE_SRC)
        return True
    except Exception:
        return False


async def fill_frame(frame: Frame, profile: dict, submit: bool) -> dict | None:
    """Call window.__AutoFill.fill() in the frame and return the result dict."""
    js = """
        ({ profile, submit }) => {
            if (!window.__AutoFill) return { skipped: true, reason: 'engine missing', url: location.href };
            return {
                url: location.href,
                ...window.__AutoFill.fill(profile, { submit: submit && window.top === window })
            };
        }
    """
    try:
        return await frame.evaluate(js, {"profile": profile, "submit": submit})
    except Exception as e:
        return {"error": str(e), "url": frame.url}


def attach_frame_listeners(target: Page | BrowserContext) -> None:
    """
    Pre-inject the engine into every new frame as soon as it attaches/navigates.
    This catches Stripe iframes that mount after first paint.
    """
    async def on_frame(frame: Frame) -> None:
        await inject_engine(frame)

    target.on("frameattached", lambda f: asyncio.create_task(on_frame(f)))
    target.on("framenavigated", lambda f: asyncio.create_task(on_frame(f)))


async def run(args: argparse.Namespace) -> int:
    store = json.loads(PROFILES_PATH.read_text(encoding="utf-8"))
    profile_key = args.profile or store.get("defaultProfile")
    profile = store["profiles"].get(profile_key) if profile_key else None
    if profile is None:
        available = ", ".join(store["profiles"].keys())
        print(f'Unknown profile "{profile_key}". Available: {available}', file=sys.stderr)
        return 1

    print(f'[autofill] profile="{profile_key}" submit={args.submit} url={args.url}')

    async with async_playwright() as pw:
        browser = await pw.chromium.launch(
            headless=not args.headful,
            channel=args.channel,
        )
        context = await browser.new_context()
        attach_frame_listeners(context)

        page = await context.new_page()
        attach_frame_listeners(page)

        await page.goto(args.url, wait_until="load")
        await page.wait_for_timeout(args.wait)

        for f in page.frames:
            await inject_engine(f)

        results = []
        for f in page.frames:
            r = await fill_frame(f, profile, submit=False)
            if r and (r.get("filled") or r.get("error") or r.get("skipped")):
                results.append(r)

        print("\n[autofill] per-frame results:")
        for r in results:
            if r.get("error"):
                print(f"  ✗ {r['url']} — {r['error']}")
            elif r.get("skipped"):
                print(f"  — {r['url']} — {r.get('reason', '')}")
            else:
                filled = ", ".join(r.get("filled", [])) or "(none)"
                missed = ", ".join(r.get("missed", [])) or "(none)"
                print(f"  ✓ {r['url']}\n    filled: {filled}\n    missed: {missed}")

        if args.submit:
            # 300ms lets the last blur-triggered validation settle before submit.
            await page.wait_for_timeout(300)
            clicked = await page.evaluate("() => window.__AutoFill && window.__AutoFill.submitForm()")
            print(f"\n[autofill] submit: {'clicked' if clicked else 'no submit button found in top frame'}")
            await page.wait_for_timeout(3000)
        elif args.headful:
            print("\n[autofill] headful mode — browser left open, Ctrl+C to exit.")
            try:
                await asyncio.Event().wait()
            except (KeyboardInterrupt, asyncio.CancelledError):
                pass

        if not args.headful:
            await browser.close()
    return 0


def main() -> None:
    try:
        sys.exit(asyncio.run(run(parse_args())))
    except KeyboardInterrupt:
        print("\ninterrupted")
        sys.exit(130)


if __name__ == "__main__":
    main()
